neuro_extras/main.py

```diff
@@ -201,6 +231,155 @@ def image() -> None:
+    if destination_url_type == UrlType.UNSUPPORTED:
+        raise ValueError(
+            f"Unsupported destination URL scheme: {destination_url.scheme}"
+        )
+
+    if source_url_type == UrlType.CLOUD and destination_url_type == UrlType.CLOUD:
+        raise ValueError(
+            "This command can't be used to copy data between cloud providers"
+        )
+    if source_url_type == UrlType.STORAGE and destination_url_type == UrlType.STORAGE:
+        raise ValueError(
+            "This command can't be used to copy data between two storage locations"
+        )
+
+    if UrlType.STORAGE in (source_url_type, destination_url_type):
+        async with neuro_api.get() as client:
+            data_copier = DataCopier(client)
+            if source_url_type == UrlType.STORAGE:
+                job = await data_copier.launch(
+                    storage_uri=source_url,
+                    src_uri=URL("/var/storage"),
+                    dst_uri=destination_url,
+                    unpack=unpack,
+                )
+            else:
+                job = await data_copier.launch(
+                    storage_uri=destination_url,
+                    src_uri=source_url,
+                    dst_uri=URL("/var/storage"),
+                    unpack=unpack,
+                )
+            while job.status == neuro_api.JobStatus.PENDING:
+                job = await client.jobs.status(job.id)
+                await asyncio.sleep(1.0)
+            async for chunk in client.jobs.monitor(job.id):
+                if not chunk:
+                    break
+                click.echo(chunk.decode(errors="ignore"), nl=False)
+            job = await client.jobs.status(job.id)
+            if job.status == neuro_api.JobStatus.FAILED:
+                logger.error("The copy job has failed due to:")
+                logger.error(f"  Reason: {job.history.reason}")
+                logger.error(f"  Description: {job.history.description}")
+                exit_code = job.history.exit_code
+                if exit_code is None:
+                    exit_code = EX_PLATFORMERROR
+                sys.exit(exit_code)
+            else:
+                logger.info("Successfully copied data")
+    else:
+        TEMP_UNPACK_DIR.mkdir(parents=True, exist_ok=True)
+        tmp_dir_name = tempfile.mkdtemp(dir=str(TEMP_UNPACK_DIR))
+        tmp_dst_url = URL.build(path=(tmp_dir_name + "/"))
+
+        await _nonstorage_cp(source_url, tmp_dst_url)
+        source_url = tmp_dst_url
+        # for clarity
+        source_url_type = UrlType.LOCAL
+
+        # at this point source is always local
+        if unpack:
+            file = Path(source_url.path)
+            if file.is_dir():
+                file = list(file.glob("*"))[0]
+            suffixes = file.suffixes
+            if suffixes[-2:] == [".tar", ".gz"] or suffixes[-1] == ".tgz":
+                command = "tar"
+                args = ["zxvf", str(file), "-C", str(destination_url)]
```

Comment:
again, destination might be a directory mounted to Storage. We should create another temp directory and then copy contents from there.

Reply:
note: with `--extract` flag, we put the original archive to destination directory, which is not needed:
```
$ ne data cp gs://cookiecutter-e2e/hello.tar.gz  /tmp/asdf -x
Running gsutil -m cp -r gs://cookiecutter-e2e/hello.tar.gz /home/ay/.neuro/tmp/tmp25zfo7bx/
Copying gs://cookiecutter-e2e/hello.tar.gz...
- [1/1 files][  177.0 B/  177.0 B] 100% Done                                    
Operation completed over 1 objects/177.0 B.                                      
Running tar zxvf /home/ay/.neuro/tmp/tmp25zfo7bx/hello.tar.gz -C /tmp/asdf
data/
data/hello.txt
Running rsync -avzh --progress /home/ay/.neuro/tmp/tmp25zfo7bx/ /tmp/asdf
sending incremental file list
./
hello.tar.gz
            177 100%    0.00kB/s    0:00:00 (xfr#1, to-chk=0/2)

sent 292 bytes  received 38 bytes  660.00 bytes/sec
total size is 177  speedup is 0.54
✓  ay@arch venv  an/38-ingestion-and-extraction  [~/github/nm/neuro-extras]
$ ls /tmp/asdf                                               
data  hello.tar.gz
```

Reply:
fixed both at 1ac716f